scipy/optimize/tests/test_optimize.py

```diff
@@ -2239,6 +2241,198 @@ def test_gh12696():
+    def func_and_grad(x):
+        assert x.size == 2
+        return optimize.rosen(x), optimize.rosen_der(x)
+
+    def wrapped_rosen(x):
+        assert x.size == 2
+        return optimize.rosen(x)
+
+    def callback(x, *args):
+        assert x.size == 2
+
+    # test example fixes one parameter of N=2, allowing us to work out
+    # the best solution with minimize_scalar
+    best_x = optimize.minimize_scalar(f, method="bounded", bounds=(0, 3.0))
+    x0 = np.array([0.5, 3.0])
+    bounds = [(0.0, 3.0), (2.0, 2.0)]
+
+    with warnings.catch_warnings(record=True):
+        # warning filter is for trust-constr
+        # UserWarning: delta_grad == 0.0.Check if the approximated
+        # function is linear.
+        warnings.simplefilter('ignore', category=UserWarning)
+
+        # the minimization with one parameter having lb == ub
+        res = optimize.minimize(
+            wrapped_rosen, x0, method=method, bounds=bounds, callback=callback
+        )
+        assert res.success
+        assert_allclose(res.x, np.r_[best_x.x, 2.0], rtol=3e-6)
+        if hasattr(res, "jac") and method in {"Powell", "L-BFGS-B", "SLSQP"}:
+            assert np.array(res.jac).size == 2
+            assert np.isnan(res.jac[1])
+
+        # check that SLSQP can deal with equal bounds if it has constraints
+        if method == "slsqp":
+            constr_best_x = optimize.minimize_scalar(
+                f, method="bounded", bounds=(2.0, 3.0)
+            )
+
+            def con(x):
+                assert len(x) == 2
+                return x[0]
+
+            def conjac(x):
+                assert len(x) == 2
+                return 1
+
+            constr = optimize.NonlinearConstraint(con, 2, 3)
+            res = optimize.minimize(
+                optimize.rosen, x0, method=method, bounds=bounds,
+                constraints=[constr]
+            )
+            assert res.success
+            assert_allclose(res.x, np.r_[constr_best_x.x, 2.0], rtol=3e-6)
+
+            constr = optimize.NonlinearConstraint(con, 2, 3, jac=conjac)
+            res = optimize.minimize(
+                optimize.rosen, x0, method=method, bounds=bounds,
+                constraints=[constr]
+            )
+            assert res.success
+            assert_allclose(res.x, np.r_[constr_best_x.x, 2.0], rtol=3e-6)
+
+        # check that minimize works if func returns func and grad.
+        if method not in ['Powell']:
+            # powell doesn't use jac
+            res = optimize.minimize(
+                func_and_grad, x0, method=method, bounds=bounds, jac=True
+            )
+            assert res.success
+            assert_allclose(res.x, np.r_[best_x.x, 2.0], rtol=3e-6)
+
+# --- Test minimize with equal upper and lower bounds --- #
+
+np.random.seed(0)
+eb_x0 = np.random.rand(4)
+eb_lb = np.array([0, 2, -1, -1])
+eb_ub = np.array([3, 2, 2, -1])
+eb_i = (eb_lb == eb_ub)
+
+
+def eb_rosen(x):
+    # assert x.size == 4
+    # assert_allclose(x[eb_i], eb_lb[eb_i])
```

Comment:
This (here and below) currently fails because TNC does not currently respect bounds where `lb==ub` (as reported [here](https://github.com/scipy/scipy/issues/14565#issuecomment-894900338)).

Reply:
Another reader may not immediately understand what `eb_rosen` is. Perhaps `_equal_bounds_rosen`, etc

Reply:
Defined all these in a function called `setup_test_equal_bounds`.